sdk/servicebus/azure-servicebus/azure/servicebus/_pyamqp/_transport.py

```diff
@@ -700,6 +711,18 @@ def __init__(
         self.ws = None
         self._http_proxy = kwargs.get("http_proxy", None)
 
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        state['socket_lock'] = None
+        state['sock'] = None
+        state['ws'] = None
+        state['_quick_recv'] = None
+        return state
+
+    def __setstate__(self, state):
+        state['socket_lock'] = Lock()
```

Comment:
They are recreated for ws, but only b/c we raise an IOError if there is no `self.ws.send_binary` attribute, and we internally set retries > 0 for IOErrors. So it closes + re-opens. For self.sock, we don't catch an AttributeError + raise IOError if `self.sock.send` doesn't exist, so it doesn't close + open w/ retry. In either case, I don't think we should be relying on retries to close + open things up, in case we make that a public pyamqp arg in the future. Will add the sock/ws creation in the setstate!

Reply:
follow-up: Trying to create a new connection in the setstate results in a 'Connection to remote host was lost error'. Since the connection is closed when we pickle, I think we need the retry loop to open the connection again. I don't think we can just call `connect()` in setstate. So, I think we'll need to try/except AttributeError in read/write in regular SSLTransport, and raise an IOError.